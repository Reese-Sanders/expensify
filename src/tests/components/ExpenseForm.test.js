import { render, screen, fireEvent } from '@testing-library/react';
import { eventWrapper } from '@testing-library/user-event/dist/utils';
import moment from 'moment';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import ExpenseForm from '../../components/ExpenseForm';
import expenses from '../fixtures/expenses';


test('should render Expense form correctly with data', () => {
  const container = render(<ExpenseForm expense={expenses[0]}/>);
  expect(container).toMatchSnapshot();
});

test('should get description input', () => {
  const container = render(<ExpenseForm />);
  const descriptionElement = container.getByPlaceholderText(/Description/i);
  expect(descriptionElement).toBeInTheDocument();
});

test('should change description input', () => {
  const container = render(<ExpenseForm />);
  const descriptionElement = container.getByPlaceholderText(/Description/i);
  fireEvent.change(descriptionElement, { target: {value: 'A new shopping spree'} });
  expect(descriptionElement.value).toBe('A new shopping spree');
});

test('should get submit button', () => {
  const container = render(<ExpenseForm />);
  const submitButton = container.getByText('Add Expense');
  expect(submitButton).toBeInTheDocument();
});

test('should submit showing an error state element', () => {
  const container = render(<ExpenseForm />);
  const submitButton = container.getByText('Add Expense');
  fireEvent.click(submitButton);
  const errorStateText = 'Description or Amount must be filled';
  const errorElement = container.getByText(errorStateText);
  expect(errorElement).toBeInTheDocument();
});

test('should change note input', () => {
  const container = render(<ExpenseForm />);
  const noteElement = container.getByPlaceholderText('Add a note for your expense (optional)');
  fireEvent.change(noteElement, { target: {value: 'A new shopping spree'} });
  expect(noteElement.value).toBe('A new shopping spree');
});


test('should change amount input', () => {
  const container = render(<ExpenseForm />);
  const amountElement = container.getByPlaceholderText('Amount');
  fireEvent.change(amountElement, { target: {value: '10.00'} });
  expect(amountElement.value).toBe('10.00');
});

test('should not change amount input', () => {
  const container = render(<ExpenseForm />);
  const amountElement = container.getByPlaceholderText('Amount');
  fireEvent.change(amountElement, { target: {value: '10,00'} });
  expect(amountElement.value).toBe('');
});

test('should call onSubit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn();
  const container = render(<ExpenseForm 
    expense={expenses[0]} 
    onSubmit={onSubmitSpy} 
    />);
  const submitButton = container.getByText('Add Expense');
  fireEvent.click(submitButton);
  expect(container.queryByText('Description or Amount must be filled')).toBeNull();
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    description: expenses[0].description,
    amount: expenses[0].amount,
    note: expenses[0].note,
    createdAt: expenses[0].createdAt
  });
});