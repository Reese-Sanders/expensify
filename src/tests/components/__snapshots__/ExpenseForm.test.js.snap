// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense form correctly with data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <form>
          <input
            placeholder="Description"
            type="text"
            value="Gum"
          />
          <input
            placeholder="Amount"
            type="text"
            value="1.95"
          />
          <div
            class="SingleDatePicker"
          >
            <div>
              <div
                class="SingleDatePickerInput"
              >
                <div
                  class="DateInput DateInput--open-down"
                >
                  <input
                    aria-describedby="DateInput__screen-reader-message-date"
                    aria-label="Date"
                    autocomplete="off"
                    class="DateInput__input needsclick"
                    id="date"
                    name="date"
                    placeholder="Date"
                    type="text"
                    value="1969-12-31"
                  />
                  <p
                    class="screen-reader-only"
                    id="DateInput__screen-reader-message-date"
                  >
                    Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
                  </p>
                  <div
                    class="DateInput__display-text DateInput__display-text--has-input"
                  >
                    12/31/1969
                  </div>
                </div>
              </div>
            </div>
          </div>
          <textarea
            placeholder="Add a note for your expense (optional)"
          />
          <button>
            Add Expense
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <form>
        <input
          placeholder="Description"
          type="text"
          value="Gum"
        />
        <input
          placeholder="Amount"
          type="text"
          value="1.95"
        />
        <div
          class="SingleDatePicker"
        >
          <div>
            <div
              class="SingleDatePickerInput"
            >
              <div
                class="DateInput DateInput--open-down"
              >
                <input
                  aria-describedby="DateInput__screen-reader-message-date"
                  aria-label="Date"
                  autocomplete="off"
                  class="DateInput__input needsclick"
                  id="date"
                  name="date"
                  placeholder="Date"
                  type="text"
                  value="1969-12-31"
                />
                <p
                  class="screen-reader-only"
                  id="DateInput__screen-reader-message-date"
                >
                  Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
                </p>
                <div
                  class="DateInput__display-text DateInput__display-text--has-input"
                >
                  12/31/1969
                </div>
              </div>
            </div>
          </div>
        </div>
        <textarea
          placeholder="Add a note for your expense (optional)"
        />
        <button>
          Add Expense
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
